librz/analysis/il_trace.c

```diff
@@ -0,0 +1,208 @@
+// SPDX-FileCopyrightText: 2021 Heersin <teablearcher@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+/**
+ * \file il_trace.c
+ * \brief new rizin IL trace implementation
+ *
+ * provide operations to new IL trace structure to record
+ * the memory changes and register changes.
+ * TODO : Should be moved to librz/il after integrations with new il
+ *      : should move the prototypes and trace structure to new header, too
+ * prototypes in <rz_analysis.h>
+ * Used by : analysis_tp.c, debug/trace.c
+**/
+
+#include <rz_util.h>
+#include <rz_analysis.h>
+
+static void free_reg_op(RzILTraceRegOp *reg_op) {
+	if (reg_op) {
+		if (reg_op->reg_name) {
+			free(reg_op->reg_name);
+		}
+		RZ_FREE(reg_op);
+	}
+}
+
+/**
+ * create and init a trace structure for an instruction at address
+ * \param addr ut64, address of instruction
+ * \return RzILTraceInstruction, trace structure of an instruction
+ */
+RZ_API RzILTraceInstruction *rz_analysis_il_trace_instruction_new(ut64 addr) {
+	RzILTraceInstruction *instruction_trace = RZ_NEW0(RzILTraceInstruction);
+	if (!instruction_trace) {
+		eprintf("Cannot create instruction trace\n");
+		return NULL;
+	}
+
+	instruction_trace->addr = addr;
+
+	instruction_trace->read_mem_ops = rz_pvector_new((RzPVectorFree)free);
+	instruction_trace->read_reg_ops = rz_pvector_new((RzPVectorFree)free_reg_op);
+	instruction_trace->write_mem_ops = rz_pvector_new((RzPVectorFree)free);
+	instruction_trace->write_reg_ops = rz_pvector_new((RzPVectorFree)free_reg_op);
+	// TODO : handle error
+	return instruction_trace;
+}
+
+/**
+ * clean an il trace
+ * \param instruction RzILTraceInstruction, trace to be cleaned
+ */
+RZ_API void rz_analysis_il_trace_instruction_free(RzILTraceInstruction *instruction) {
+	if (instruction->write_reg_ops) {
+		rz_pvector_free(instruction->write_reg_ops);
+		instruction->write_reg_ops = NULL;
+	}
+
+	if (instruction->read_reg_ops) {
+		rz_pvector_free(instruction->read_reg_ops);
+		instruction->read_reg_ops = NULL;
+	}
+
+	if (instruction->write_mem_ops) {
+		rz_pvector_free(instruction->write_mem_ops);
+		instruction->write_mem_ops = NULL;
+	}
+
+	if (instruction->read_mem_ops) {
+		rz_pvector_free(instruction->read_mem_ops);
+		instruction->read_mem_ops = NULL;
+	}
+
+	RZ_FREE(instruction);
+}
+
+/**
+ * add memory change to an instruction trace
+ * \param trace RzILTraceInstruction *, trace of instruction which triggers a memory change
+ * \param mem RzILTraceMemOp *, info of memory change
+ */
+RZ_API void rz_analysis_il_trace_add_mem(RzILTraceInstruction *trace, RzILTraceMemOp *mem) {
+	if (!trace || !mem) {
+		return;
+	}
+
+	int is_write = mem->behavior;
+	if (rz_analysis_il_mem_trace_contains(trace, mem->addr, is_write)) {
+		return;
+	}
+
+	if (is_write) {
```

Comment:
it might make sense to use something like:
```c
switch (op_type) {
case RZ_IL_TRACE_OP_READ:
  rz_pvector_push(trace->read_mem_ops, mem);
  trace->stats |= TRACE_INS_HAS_MEM_R;
case RZ_IL_TRACE_OP_WRITE:
  rz_pvector_push(trace->write_mem_ops, mem);
  trace->stats |= TRACE_INS_HAS_MEM_W;
default:
  rz_warn_if_reached();
  break;
}
```

This way if you ever add some other values to the op_type enum you will immediately see you have to touch this code.

Reply:
Same for the other places ofc.